hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/TarExtractor.java

```diff
@@ -0,0 +1,181 @@
+      parentDir = outputDir; // Handle null parent case
+    }
+    Path stagingDir = parentDir.resolve(stagingDirName);
+
+    Files.createDirectories(stagingDir); // Ensure staging directory exists
+
+    List<Future<Void>> futures = new ArrayList<>();
+
+    try (TarArchiveInputStream tarInput = new TarArchiveInputStream(tarStream)) {
+      TarArchiveEntry entry;
+      while ((entry = tarInput.getNextTarEntry()) != null) {
+        if (!entry.isDirectory()) {
+          byte[] fileData = readEntryData(tarInput, entry.getSize());
+
+          // Submit extraction as a task
+          TarArchiveEntry finalEntry = entry;
+          futures.add(executor.submit(() -> {
+            writeFile(stagingDir, finalEntry.getName(), fileData);
+            return null;
+          }));
+        } else {
+          File dir = new File(stagingDir.toFile(), entry.getName());
+          if (!dir.exists() && !dir.mkdirs()) {
+            throw new IOException("Failed to create directory: " + dir);
+          }
+        }
+      }
+    }
+
+    // Wait for all tasks to complete
+    for (Future<Void> future : futures) {
+      future.get();
+    }
+
+    // Move staging to outputDir atomically
+    if (Files.exists(outputDir)) {
+      FileUtils.deleteDirectory(outputDir.toFile()); // Clean old data
+    }
+    try {
+      Files.move(stagingDir, outputDir, StandardCopyOption.ATOMIC_MOVE, StandardCopyOption.REPLACE_EXISTING);
+    } catch (IOException e) {
+      LOG.warn("Atomic move of staging dir : {} to {} failed.", stagingDir, outputDir, e);
+    }
+
+    LOG.info("Tar extraction completed and moved from staging to: {}", outputDir);
+  }
+
+  private byte[] readEntryData(TarArchiveInputStream tarInput, long size) throws IOException {
+    ByteArrayOutputStream buffer = new ByteArrayOutputStream();
+    byte[] fileData = new byte[(int) size];
+    int bytesRead;
+    while (size > 0 && (bytesRead = tarInput.read(fileData, 0, (int) Math.min(fileData.length, size))) != -1) {
+      buffer.write(fileData, 0, bytesRead);
+      size -= bytesRead;
+    }
+    return buffer.toByteArray();
+  }
+
+  private void writeFile(Path outputDir, String fileName, byte[] fileData) {
+    try {
+      File outputFile = new File(outputDir.toFile(), fileName);
+      Path parentDir = outputFile.toPath().getParent();
+      if (parentDir != null && !Files.exists(parentDir)) {
+        Files.createDirectories(parentDir);
+      }
+
+      try (InputStream fis = new ByteArrayInputStream(fileData);
+           OutputStream fos = Files.newOutputStream(outputFile.toPath())) {
+        byte[] buffer = new byte[8192];  // Use a buffer for efficient reading/writing
+        int bytesRead;
+        while ((bytesRead = fis.read(buffer)) != -1) {
+          fos.write(buffer, 0, bytesRead);
+        }
+      }
+    } catch (IOException e) {
+      throw new RuntimeException("Error writing file: " + fileName, e);
+    }
+  }
+
+  public void start() {
+    this.executor =
```

Comment:
start()/stop() can be safe from leaking the threads in threadpool

Reply:
Ok done.